### PR TITLE
Prototype: $count processed by WebApi implemented QueryExecutor.

### PR DESCRIPTION
**Pull request by rayao**
#323 
I think it's basically like this, don't know whether meets your thoughts.

### CODE REVIEW

**Comment by lewischeng-ms** [comment]
@chinadragon0515 Please help review Ray's design proposal together. Thanks.

**Comment by rayao** [comment]
@lewischeng-ms @chinadragon0515 No hurry, and it's a prototype just to share my understanding of your intention.
Enjoy your holiday Lewis, it'll be a wonderful trip I believe. :)

**Comment by lewischeng-ms** [comment]
Thanks @rayao :-)

File: src/Microsoft.Restier.WebApi/RestierController.cs [hunk 1]
```diff
@@ -76,15 +76,11 @@ public IApi Api
             }
 
             IQueryable queryable = this.GetQuery();
-            QueryRequest queryRequest = new QueryRequest(queryable, this.includeTotalCount)
+            QueryRequest queryRequest = new QueryRequest(queryable, false)
             {
                 ShouldReturnCount = this.shouldReturnCount
             };
             QueryResult queryResult = await Api.QueryAsync(queryRequest, cancellationToken);
```
Comment (chinadragon0515) [inline src/Microsoft.Restier.WebApi/RestierController.cs]: There maybe a race condition here? When queryAsync is executed, the context QueryIncludeTotalCount is not set yet, and cause QueryExecutorWebApi does not call count query?

File: src/Microsoft.Restier.WebApi/Query/QueryExecutorWebApi.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Web.OData.Extensions;
+using Microsoft.Restier.Core.Query;
+
+namespace Microsoft.Restier.WebApi.Query
+{
+    internal class QueryExecutorWebApi : IQueryExecutor
```
Comment (chinadragon0515) [inline src/Microsoft.Restier.WebApi/Query/QueryExecutorWebApi.cs]: rename to WebApiQueryExecutor?

File: src/Microsoft.Restier.WebApi/Query/QueryExecutorWebApi.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            IQueryable<TElement> query,
+            CancellationToken cancellationToken)
+        {
+            var queryContext = context.ApiContext.GetApiService<WebApiContext>();
+            if (queryContext.QueryIncludeTotalCount == true)
+            {
+                var countQuery = ExpressionHelpers.GetCountableQuery(query);
+                var expression = countQuery.Expression;
+                expression = ExpressionHelpers.Count(expression, countQuery.ElementType);
+                var result = await ExecuteSingleAsync<long>(context, countQuery, expression, cancellationToken);
```
Comment (chinadragon0515) [inline src/Microsoft.Restier.WebApi/Query/QueryExecutorWebApi.cs]: General comment: I checked the code, seems ExecuteSingleAsync key logic will be countQuery.provider.execute(expression), then question will be whether we need two parameters here (query and expression) or rename query to QueryProvider?

File: src/Microsoft.Restier.WebApi/Query/QueryExecutorWebApi.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            QueryContext context,
+            IQueryable<TElement> query,
+            CancellationToken cancellationToken)
+        {
+            var queryContext = context.ApiContext.GetApiService<WebApiContext>();
+            if (queryContext.QueryIncludeTotalCount == true)
+            {
+                var countQuery = ExpressionHelpers.GetCountableQuery(query);
+                var expression = countQuery.Expression;
+                expression = ExpressionHelpers.Count(expression, countQuery.ElementType);
```
Comment (chinadragon0515) [inline src/Microsoft.Restier.WebApi/Query/QueryExecutorWebApi.cs]: What's the major purpose of this call? seems we make it generic with specific type, and make it as call, the expression is already object type which is enough for count, also I have not tested whether countQuery.ElementType can get more specific type here besides object.

File: src/Microsoft.Restier.WebApi/WebApiContext.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace Microsoft.Restier.WebApi
+{
+    internal class WebApiContext
+    {
+        public HttpRequestMessage Request { get; set; }
+
+        public bool? QueryIncludeTotalCount { get; set; }
+    }
+}
```
Comment (chinadragon0515) [inline src/Microsoft.Restier.WebApi/WebApiContext.cs]: In general, I prefer we do not introduce a new class with only one additional properties, can we just use request.Properties to store one more attribute to indicate need total count returned?

**Comment by rayao** [comment]
A complete PR sent as #329, close this.
Sorry @chinadragon0515 for wasting your time to review.

**Pull request closed without merging**